### PR TITLE
[refactor] Add SNode::GradInfoProvider to isolate SNode from Expr

### PR DESCRIPTION
**Pull request by k-ye**
Related issue = #2196 

This PR is still about decoupling `SNode` from `taichi/program`. The major issue here is around its member variable `expr`. Because `Expr` is part of the frontend and relies on `taichi/program`, we have to move it out from `SNode`:

* Moved `place()` and `lazy_grad()` from `SNode` itself to a separate file `taichi/program/snode_expr_utils.h/cpp` (i cannot think of a better file name...). Other functions in `SNode` do not depend on `expr`.
* Created a `SNode::GradInfoProvider` abstract class. This is essentially a wrapper around `GlobalVariableExpression`. From what I can tell, `SNode::expr` is mostly used to provide gradient information, e.g. `is_primal()`, `has_grad()`, etc. The actual implementation, `GradInfoImpl`, is in `taichi/program/snode_expr_utils.cpp`.
* Created a map from `SNode*` to `GlobalVariableExpression`. The map is stored inside `Program`. While it is possible to retrieve `GlobalVariableExpression` from `GradInfoImpl`, that involves casting `GradInfoProvider` to `GradInfoImpl` and kind of exposes the implementation detail. I'd like to make `GradInfoProvider` have a single responsibility.

[[Click here for the format server]](http://kun.csail.mit.edu:31415/)

----

### CODE REVIEW

File: taichi/struct/struct.cpp [hunk 1]
```diff
@@ -95,9 +96,6 @@ void StructCompiler::infer_snode_properties(SNode &snode) {
     }
   }
 
-  if (snode.expr.expr)
```
Comment (k-ye) [inline taichi/struct/struct.cpp:98]: This is moved to `Program::materialize_snode_expr_attributes()`

File: taichi/program/snode_expr_utils.cpp [hunk 2]
```diff
@@ -0,0 +1,110 @@
...
+    // never directly place to root
+    auto &ds = parent->dense(std::vector<Index>(), {});
+    place_child(expr_arg, offset, &ds, snode_to_exprs);
+  } else {
+    TI_ASSERT(expr_arg->is<GlobalVariableExpression>());
+    auto glb_var_expr = expr_arg->cast<GlobalVariableExpression>();
+    TI_ERROR_IF(glb_var_expr->snode != nullptr,
+                "This variable has been placed.");
+    SNode *new_exp_snode = nullptr;
+    if (auto cft = glb_var_expr->dt->cast<CustomFloatType>()) {
```
Comment (k-ye) [inline taichi/program/snode_expr_utils.cpp:45]: This part (around quant types) may be OK to be left inside `SNode` itself. But we can change that later if necessary..

Comment (yuanming-hu) [reply taichi/program/snode_expr_utils.cpp:45]: Sounds good. Let's decide after the quant API/impl materializes...

**Review by yuanming-hu — APPROVED** [review]
Awesome! I like the `GradInfoProvider` class.

File: taichi/program/program.cpp [hunk 3]
```diff
@@ -633,13 +634,13 @@ Arch Program::get_snode_accessor_arch() {
...
   auto kernel_name = fmt::format("snode_reader_{}", snode->id);
-  auto &ker = kernel([snode] {
+  auto &ker = kernel([snode, this] {
     ExprGroup indices;
     for (int i = 0; i < snode->num_active_indices; i++) {
       indices.push_back(Expr::make<ArgLoadExpression>(i, PrimitiveType::i32));
     }
     auto ret = Stmt::make<FrontendKernelReturnStmt>(
-        load_if_ptr((snode->expr)[indices]));
+        load_if_ptr(Expr(snode_to_glb_var_exprs_.at(snode))[indices]));
```
Comment (xumingkuan) [inline taichi/program/program.cpp:643]: I wonder why we need this `Expr()`? (same question at line 663)

**Review by xumingkuan — APPROVED** [review]
LGTM!

**Pull request merged**